feat: targeted / untargeted spectra matching

pcolaianni opened the pull request:
Please ignore the long list of commits.
Just check the code for `targetedMatching()` and `untargetedMatching()`.
I'll take care of cleaning up the history when the code will be ready to be proposed upstream.

Is this close to what we need?
I am not sure if the naming of the two methods really explains what the code does.
The fact that the untargeted procedure actually calls the `targetedMatching()` is at least confusing.
Basically, the untargeted version simply creates a `FeatureMap` from the input (and picked) spectra (because these spectra did not go through a process of annotation, peak picking, scoring, selection, extraction), and then the same job as `targetedMatching()` is needed... therefore I use it within `untargetedMatching()`.

`targetedMatching()`'s input spectra would be those outputted from `extractSpectra()` (or from `untargetedMatching()`).

pcolaianni committed:
feat: implement MSP format parser for TargetedSpectraExtractor::matchSpectrum()

pcolaianni committed:
feat: matchSpectrum() prototype

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
docs: update docs (review)

feat: add top_matches_to_report class parameter

chore: remove <boost/unordered_map.hpp> include

chore: use <regex> instead of boost version

test: isolate matchSpectrum tests from others (ie. contants, params, etc)

Tests currently fail because of some metadata missing from experiment file

pcolaianni committed:
chore: Rename MSPTSEFile to TseMSPFile

pcolaianni committed:
chore: 18 extracted spectra

pcolaianni committed:
fix: refine params (gauss width and snr)

pcolaianni committed:
feat: avoid adding duplicates in msp library

feat: add "Num Peaks" metadata

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
feat: implement `pushParsedInfoToNamedDataArray()`

pcolaianni committed:
docs: Add small comments to new method's if/else

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
feat: implement addSpectrumToExperiment()

feat: implement getStringDataArrayByName()

feat: make loaded_spectra_names_ a class member

test: add friend class to test or use private/protected methods

test: print synonyms

pcolaianni committed:
style: rename TseMSPFile to MSPMetaboFile

pcolaianni committed:
feat: parse generic metadata

pcolaianni committed:
fix: set `adding_spectrum` to false when a duplicate is found

feat: implement check on Name

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
feat: ensure Comments metadatum is present

pcolaianni committed:
test: verify that absent Comments is an empty string

pcolaianni committed:
feat: implement class params bin_size and peak_spread

test: display results differently

pcolaianni committed:
style: rename experiment to library in matchSpectrum()

pcolaianni committed:
fix: check that top_matches_to_report_ is not higher than the actual number of spectra available

pcolaianni committed:
fix: reset loaded_spectra_names before loading a file

pcolaianni committed:
test: update library size, remove CAS# cout

pcolaianni committed:
test: implement matchIsValid()

pcolaianni committed:
Merge branch 'fix/annotateSpectra_no_break' into feature/TSE_MatchSpectrum

pcolaianni committed:
feat: Move peaks parsing up to increase speed

docs: add LOG_INFO messages to show during parse

pcolaianni committed:
feat: add binned spectra library as class-member to speedup matching

test: make expected_ global

test: add createTabularSpectra() to easily compare matches

pcolaianni committed:
feat: increase speed of parser (replace vector with set)

pcolaianni committed:
style: rename experiment to library

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
Merge remote-tracking branch 'upstream/featzre/binned_offset' into feature/TSE_MatchSpectrum

pcolaianni committed:
feat: implement bin_offset

test: exceptions

pcolaianni committed:
Merge remote-tracking branch 'upstream/develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
fix: merge conflict

pcolaianni committed:
docs: bs_library_ class member (spectra library)

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
style: rename experiment to library

pcolaianni committed:
feat: add min_match_score class parameter

feat: implement extractBinnedSpectrum()

feat: add Match struct, to be used

pcolaianni committed:
feat: simplify extractBinnedSpectrum interface (returns a reference instead of iterator)

pcolaianni committed:
style: rename min_score class parameter to min_select_score

pcolaianni committed:
feat: use Match struct instead of pair<string,double>

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
refactor: replace insert() with emplace()

pcolaianni committed:
feat: prepare smaller gcms experiment

pcolaianni committed:
test: remove unnecessary features computation

pcolaianni committed:
feat: add MSPMetaboFile .pxd

pcolaianni committed:
refactor: remove unnecessary regexp and related "else/if"s

pcolaianni committed:
docs: decrease info shown (every 20k spectra loaded)

pcolaianni committed:
style: remove unnecessary lines, improve comments

pcolaianni committed:
docs: fix var name in argument list

pcolaianni committed:
feat: MSSpectrum methods to return const/mutable reference to a named data array

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
Merge branch 'feature/MSSpectrum_getTYPEDataArrayByName' into feature/TSE_MatchSpectrum

pcolaianni committed:
refactor: update getStringDataArrayByName() calls (now part of MSSpectrum)

pcolaianni committed:
refactor: remove unnecessary map use from matchSpectrum()

pcolaianni committed:
style: rename bs_library_name to bs_key

docs: add description of bs_key

pcolaianni committed:
feat: 1st version of targeted and untargeted matching methods

pcolaianni committed:
docs: improve logged info to console

pcolaianni committed:
refactor: remove dependency on SpectrumHelper.h

pcolaianni committed:
test: update TSE tests for untargetedMatching

pcolaianni committed:
feat: support another peak format: x1:y1 x2:y2 ...

refactor: avoid try/catch block

pcolaianni committed:
refactor: simplify regex

pcolaianni committed:
docs: add documentation to private methods

pcolaianni committed:
fix: do not add name to names library if num peaks check throws exception

test: add tests for private methods

pcolaianni committed:
refactor: replace bool with metavalue for validating a spectrum

pcolaianni committed:
docs: fix words

pcolaianni committed:
Merge branch 'feature/MSPMetaboFile' into feature/TSE_MatchSpectrum

pcolaianni committed:
Merge remote-tracking branch 'upstream/develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
Merge branch 'feature/MSPMetaboFile' into feature/TSE_MatchSpectrum

pcolaianni committed:
test: refining tests

pcolaianni committed:
Merge branch 'feature/MSPMetaboFile' into feature/matchSpectrum

pcolaianni committed:
feat: implement matchSpectrum()

style: rename min_score_ to min_select_score_

pcolaianni committed:
fix: remove cmake warning

pcolaianni committed:
feat: update pxd

pcolaianni committed:
test: add data files (experiment, target list, library)

pcolaianni committed:
test: update filename

pcolaianni committed:
Merge branch 'develop' into feature/matchSpectrum

pcolaianni committed:
test: fix accessing unexistant match

pcolaianni committed:
Merge branch 'feature/MSPMetaboFile' into feature/matchSpectrum

pcolaianni committed:
refactor: Comparator class

pcolaianni committed:
Merge branch 'develop' into feature/matchSpectrum

pcolaianni committed:
refactor: replace variadic function with "options" map

refactor: remove operator(), replace it with generateScores()

pcolaianni committed:
test: fix data filename

pcolaianni committed:
fix: make sure private params have a value; check before accessing map

pcolaianni committed:
fix: pass arguments by reference in .pxd file

pcolaianni committed:
test: verify the score but not the match name

pcolaianni committed:
Merge branch 'develop' into feature/targeted_untargeted_matching

pcolaianni committed:
Merge branch 'feature/matchSpectrum' into feature/targeted_untargeted_matching

pcolaianni committed:
refactor: save library inside Comparator, do not pass it to matchSpectrum()

pcolaianni committed:
fix: implement Comparator in pxd

pcolaianni committed:
Merge branch 'feature/matchSpectrum' into feature/targeted_untargeted_matching

pcolaianni committed:
fix: do not pass library around

pcolaianni committed:
fix: assume one metadatum per line

Because char ";" is not universally used as separator for metadata

pcolaianni committed:
Merge branch 'fix/MSPGenericFile_one_metadatum_per_line' into feature/targeted_untargeted_matching

pcolaianni committed:
feat: implement targeted and untargeted matching

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:
```diff
@@ -473,4 +522,127 @@ namespace OpenMS
+      }
+      else
+      {
+        no_matches_idx.push_back(i);
```
Probably we should insert an empty data if no matches are found.  I have found problems when writing out the featureMap in .tsv/.csv format when metaValues differed between the features.

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:
```diff
@@ -473,4 +522,127 @@ namespace OpenMS
+      features.push_back(feature);
+    }
+
+    targetedMatching(picked, cmp, features);
```
This looks fine to me.  We will need to test it out to see how it works in practice before we can say more.

dmccloskey commented on src/openms/source/FORMAT/MSPGenericFile.cpp:
```diff
@@ -126,21 +127,25 @@ namespace OpenMS
         spectrum.setName( String(m[1]) );
         spectrum.setMetaValue("is_valid", 1);
       }
+      // Specific case of NIST's exported msp
```
I wonder how many other "specific" cases there are for the other databases.  This maybe a detail worth documenting in the docstrings for future developers (or even us) if we have to add in another "specific" case at some point in time.

pcolaianni replied:
Yeah, the absence of precise rules for how a msp file should be formatted is not cool.

I explain the change in better words in the following [FIX] PR to upstream: https://github.com/OpenMS/OpenMS/pull/3574

dmccloskey reviewed (approved):
I think this is spot on to what we had spoken about.

In regards to the name of "targetedMatching" and it's use in "untargetedMatching", I think that is fine.

pcolaianni committed:
Merge branch 'develop' into feature/targeted_untargeted_matching

pcolaianni committed:
docs: add doxygen documentation

feat: add default values for missing matches

pcolaianni commented:
@dmccloskey 

Added a change that deals with "missing match" case, as recommended in a comment.

Also added doxygen documentation.

How do those two points look?

dmccloskey reviewed (approved):
This looks good to me.  I think it is ready for a PR to OpenMS.
